Forms: Rating Field: fix filled icons and stacking in Twenty Sixteen; switch to Stars/Hearts variations

lezama opened the pull request:
Fixes #

## Proposed changes:
- Replace useEffect-based default handling with a BlockContextProvider to stop the rating field from marking the post as dirty when interacting in the editor.
- Introduce block variations for styles (Stars, Hearts) and remove the custom Rating toolbar.
- Normalize markup and classes to the `jetpack-field-rating__*` pattern; update both the editor and PHP-rendered markup to use:
  - `jetpack-field-rating__options` (container)
  - `jetpack-field-rating__option` (each item)
  - `jetpack-field-rating__input` (radio)
  - `jetpack-field-rating__label` (label)
  - `jetpack-field-rating__icon` (SVG)
- Switch SVG paths to outlines (`fill: none; stroke: currentColor`) and add CSS that fills icons only on hover or when selected, including previous-sibling fill for earlier options.
- Fix vertical stacking by applying flex layout to `jetpack-field-rating__options`.
- Add a scoped font-size guard on labels to prevent the Twenty Sixteen theme from shrinking icons.

### Other information:
- [ ] Have you written new tests for your changes, if applicable? No new tests needed; styling/markup and editor behavior fix
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
Fixes two issues in Twenty Sixteen: rating icons appeared filled by default and stacked vertically. Also resolves the rating block causing an unsaved (dirty) post state in the editor.

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
- Activate Twenty Sixteen.
- Editor:
  1. Create a new post/page.
  2. Add a Form block → add a Rating field.
  3. Switch between Stars and Hearts variations.
  4. Hover/choose a rating:
     - Current and all previous icons fill.
     - Icons remain outline by default.
     - Items are laid out horizontally.
  5. Change Max value; verify the correct number of icons render.
  6. Interact with the rating field (hover/click to preview default). Confirm the post is not marked as having unsaved changes unless you explicitly change a setting.
- Frontend:
  1. View the post.
  2. Verify icons are outlines by default.
  3. Hover/select:
     - Current and previous icons fill correctly.
     - Items are laid out horizontally.
  4. Confirm label/icon sizes look correct (font-size guard works under Twenty Sixteen).

- Optional: Repeat with another modern block theme to ensure no regressions.

lezama committed:
Fix rating field icons showing as filled in Twenty Sixteen theme

- Change SVG icons from fill='currentColor' to fill='none' for unfilled state
- Add CSS rules to fill icons only when selected or hovered
- Update both PHP rendered icons and React editor icons to match
- Ensures hearts/stars show as outlines by default, filled when selected

lezama committed:
Fix rating icons displaying vertically in Twenty Sixteen theme

- Add flex display to parent container for horizontal layout
- Use gap for consistent spacing between icons
- Set flex-shrink: 0 to prevent icons from squishing
- Change label display to inline-flex for better alignment

lezama committed:
Forms: Refactor rating field for consistent editor/frontend rendering

lezama committed:
Fix hearts/stars icon detection using correct attribute

lezama committed:
Replace star icon with custom SVG in rating toolbar

Replaces the imported star icon with a custom SVG star icon to better match the size and style of the custom heart icon in the rating toolbar component.

enejb commented:
There are some nice fixes but I also found some issues. 

1. The hover effect doesn't work well when the default value is set to 0. (The previous stars don't highlight) 
2. The label doesn't pick up the default styles. 
3. The colour is not set in the editor. But it does show up on the front end of the site as expected.

lezama committed:
Improve rating input styles for checked and hover states

lezama committed:
Refactor rating field to use block variations and color support

Replaces the previous styles array for rating field with a new variations system, enabling block variations for stars and hearts. Adds support for WordPress color classes and styles in the rating input, updates SCSS for color and font size support, and refines class handling in both JS and PHP render output. Removes unused toolbar and style logic from the edit component.

lezama committed:
Refactor rating input styles and SVG attributes

lezama committed:
Refactor rating input styles for font and color support

Simplifies the input-rating SCSS by removing custom font-size and color variable logic, instead relying on core font-size and color inheritance. Ensures icons scale with font-size and inherit color, improving compatibility with theme styles.

lezama committed:
cleanup

lezama committed:
more cleanup

lezama committed:
Refactor rating input option class and update styles

Replaces the 'contact-form-field wp-block-jetpack-option' class with 'jetpack-field-rating__option' in both the React and PHP implementations of the rating input. Updates the SCSS to improve layout, accessibility, and visual states for the rating input, consolidating and clarifying the styling rules.

lezama committed:
Rename rating options class for consistency

Updated the CSS class from 'jetpack-rating-options' to 'jetpack-field-rating__options' in JavaScript, SCSS, and PHP to ensure consistent naming across the input rating block implementation.

simison commented on projects/packages/forms/src/blocks/field-rating/variations.js:
```diff
@@ -0,0 +1,69 @@
+	},
+	{
+		name: 'hearts',
+		title: __( 'Hearts Rating', 'jetpack-forms' ),
```
```suggestion
		title: __( 'Hearts rating field', 'jetpack-forms' ),
```

Should we call the variations "fields" too so that they appear in pickers and elsewhere as form fields?

That will help differentiate from other ratings things:

<img width="239" height="198" alt="Screenshot 2025-08-08 at 10 54 14" src="https://github.com/user-attachments/assets/3d7d6d46-151b-43d3-a04f-98d1ce362279" />

simison commented on projects/packages/forms/src/blocks/input-rating/style.scss:
```diff
@@ -1,72 +1,51 @@
-			height: 1em;
-			display: block;
+		/* Accessibly hide the actual radios */
+		input[type="radio"].jetpack-field-rating__input {
```
We have `visually-hidden` class already in `grunion.css` so alternative would be to use that and save some bytes.

simison commented on projects/packages/forms/src/blocks/input-rating/style.scss:
```diff
@@ -1,72 +1,51 @@
-		color: inherit;
-		cursor: pointer;
+		// Filled states: current (checked/hover)
+		:is(input[type="radio"].jetpack-field-rating__input:checked + .jetpack-field-rating__label, .jetpack-field-rating__label:hover) .jetpack-field-rating__icon path {
```
Does this work without `&`? — `&:is()`

simison commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -2131,40 +2131,47 @@ private function render_rating_field( $id, $label, $value, $class, $required, $r
+				'<div class="jetpack-field-rating__option">
 					<input
-						class="jetpack-field-rating__input"
+						id="%7$s"
```
Nit but I'd prolly arrange these attributes in order 1-2-3-etc just for readability.

simison commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -2131,40 +2131,47 @@ private function render_rating_field( $id, $label, $value, $class, $required, $r
+					</label>
+				</div>',
 				checked( $i, $initial_rating, false ),
 				$required ? 'required aria-required="true"' : '',
```
Do we need `aria-required` if we already have `required`? 🤔

simison commented on projects/packages/forms/src/blocks/input-rating/edit.js:
```diff
@@ -1,55 +1,84 @@
+	for ( let i = 1; i <= max; i++ ) {
+		const radioId = `rating-${ clientId }-${ i }`;
+		ratingOptions.push(
+			<div key={ i } className="jetpack-field-rating__option">
```
Bummer to use index as key but I don't really see alternative. 🤔 

https://react.dev/learn/rendering-lists#why-does-react-need-keys

simison commented:
Colour gets set via theme's colours (purple), but not via custom colours (pink):

<img width="1204" height="555" alt="Screenshot 2025-08-08 at 11 00 52" src="https://github.com/user-attachments/assets/7f37c670-d601-4aa5-b005-78dcd5d9fe6e" />

Variation picker doesn't show currently active selection but otherwise works well:

<img width="283" height="145" alt="Screenshot 2025-08-08 at 11 01 04" src="https://github.com/user-attachments/assets/0b608014-7f9e-4ec1-94e7-25a9e00858a3" />

simison reviewed (dismissed):
Let's get this merged, perhaps after some minor field renames, and continue in follow-ups on other fixes?

simison commented on projects/packages/forms/src/blocks/field-rating/variations.js:
```diff
@@ -0,0 +1,69 @@
+						strokeWidth="2"
+						strokeLinejoin="round"
+					/>
+				</SVG>
```
```suggestion
				<Path
					d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.62L12 2 9.19 8.62 2 9.24l5.46 4.73L5.82 21z"
					fill="none"
					stroke="currentColor"
					strokeWidth="2"
					strokeLinejoin="round"
				/>
```

I see this working fine without `<svg>` elsewhere with `renderMaterialIcon()` so worth try.

lezama committed:
Refactor RatingInputEdit to remove unused attributes

Removed the unused 'attributes' prop and related logic from the RatingInputEdit component. The component now relies solely on context for configuration, simplifying its interface and reducing unnecessary code.

lezama committed:
Add CSS variable for rating star color

lezama committed:
Refactor rating input styles for clarity and hover logic

lezama committed:
Refactor rating field to use iconStyle attribute

lezama committed:
Update projects/packages/forms/src/blocks/field-rating/variations.js

Co-authored-by: Mikael Korpela <mikael@ihminen.org>

lezama committed:
Refactor rating field icon rendering and radio markup

Simplifies the icon rendering in field-rating variations by removing unnecessary SVG wrappers. Updates the contact form rating field radio input markup to use clearer and more consistent sprintf argument ordering, improving readability and maintainability.

lezama committed:
Remove shared rating block JS and update rating field PHP

Deleted unused shared JavaScript files for the rating block, including the toolbar, hook, and helpers. Updated PHP to support a new 'iconstyle' attribute for rating fields, ensuring compatibility with shortcode processing and proper icon rendering. Also set a default 'iconStyle' in the contact form plugin when rendering rating fields.

enejb reviewed (dismissed):
Works well in my testing!
